microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/commons/utils/PropertyValueExpander.java

```diff
@@ -0,0 +1,241 @@
+/*******************************************************************************
+* Copyright (c) 2022 Red Hat Inc. and others.
+*
+* This program and the accompanying materials are made available under the
+* terms of the Eclipse Public License v. 2.0 which is available at
+* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+* which is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*
+* Contributors:
+*     Red Hat Inc. - initial API and implementation
+*******************************************************************************/
+package org.eclipse.lsp4mp.commons.utils;
+
+import java.text.MessageFormat;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.logging.Logger;
+
+import com.google.common.graph.EndpointPair;
+import com.google.common.graph.Graph;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.Graphs;
+import com.google.common.graph.MutableGraph;
+import com.google.common.graph.Traverser;
+
+import io.smallrye.common.expression.Expression;
+import io.smallrye.common.expression.Expression.Flag;
+
+/**
+ * Expands property expressions to a final resolved value, while avoiding a few
+ * pitfalls.
+ * 
+ * Checks for cycles before attempting expansion, and mitigates OOM due to
+ * Billion Laughs by counting the number of variable references.
+ * 
+ * @author datho7561
+ */
+public class PropertyValueExpander {
+
+	private static final Logger LOGGER = Logger.getLogger(PropertyValueExpander.class.getName());
+	private static final long REFERENCE_UPPER_BOUND = 1_000_000;
+
+	private final PropertyGraph propertyGraph;
+	private final Map<String, String> properties;
+	private final Map<String, ResolvedPropertyValueInformation> resolved;
+
+	public PropertyValueExpander(Map<String, String> properties) {
```

Comment:
Having PropertyValueExpander in commons is very nice but it requires to build a temporary Map to consume it. For MP LS, it means that this temporary Map will be build on each key stroke which is not good for performance.

Instead of using Map<String, String> properties, I will try to introduce a `IConfigSourcePropertiesProvider` which will give the capability to get information that you need (keys, values). IConfigSourcePropertiesProvider implementation is a wrapper which be implemented internally with :

 * list of IConfigSource (on JDT side)
 * a PropertiesModel on MP LS side (that we will change into a list of PropertiesModel in the future).

Reply:
Let me know what you think of my implementation.